Rust cleanup

emesare opened the pull request:
1. Remove architecture trait bound on LowLevelILFunctions (this then impacts other types)
2. Remove infallible BnString conversions
3. Misc cleanup

For 1 I think this is pretty easy migration for users, they will just need remove the extra bound, build errors are descriptive enough.

For 2 this just means that users are more responsible for handling invalid UTF-8 which is the correct call in the cases where we cant already just make that assumption, or that a lossy conversion is acceptable.

I need to move on to some other tasks so this PR will likely not grow, I am going to mark this ready for review.

CC: @bdash @rbran 

Related: https://github.com/Vector35/binaryninja-api/issues/6784 https://github.com/Vector35/binaryninja-api/issues/6778

rbran committed:
[Rust] Implement `LanguageRepresentation` and `LineFormatter`

rbran committed:
[Rust] Implement custom interactive handler

emesare committed:
[Rust] Pretty print LLIL sub expressions

emesare committed:
[Rust] Remove Architecture trait bound on LLIL related structures

emesare committed:
[Rust] Return `String` instead of `BnString` for cases where lossy conversion can be tolerated

Still need to go and audit all usage, but realistically the most important places to give the user control are with symbols, where the data can come from non utf8 sources

This is still incomplete, I just looked for usage of -> BnString so any other variant was omitted.

emesare committed:
[Rust] Remove unused file

emesare committed:
[Rust] More cleanup

emesare committed:
[Rust] Misc clippy lints

bdash commented on rust/src/string.rs:
```diff
@@ -69,11 +69,26 @@ impl BnString {
 
+    /// Take an owned core string and convert it to [`String`].
+    ///
+    /// This expects the passed raw string to be owned, as in, freed by us.
```
This should probably document the behavior when the input contains non-UTF-8 characters.

bdash commented on rust/src/string.rs:
```diff
@@ -69,11 +69,26 @@ impl BnString {
+    ///
+    /// This expects the passed raw string to be owned, as in, freed by us.
+    pub unsafe fn into_string(raw: *mut c_char) -> String {
+        Self::from_raw(raw).to_string()
```
This confused me for a bit. Is there a reason this calls `to_string` rather than directly calling `to_string_lossy`? IMO the function name should also indicate that this is a lossy conversion.

It looks like `to_string` is available on `BnString` due to the blanket impl of `ToString` for any type implementing `Display`.  `BnString`'s `Display` impl delegates to `to_string_lossy`.

Having `to_string` be available but effectively aliased to `to_string_lossy` seems surprising. It might be better to follow `CString`'s example and not implement `Display` or `ToString`.

emesare replied:
Not having the display impl is kinda annoying, maybe we could get away with demoting it to a debug impl that is lossy?

bdash replied:
The `Debug` impl is sufficient for logging with `{:?}`. There's a lot of types that require that rather than the `{}` that `Display` allows. It could be a little annoying to adjust to, but it still seems better than silently doing a lossy conversion.

emesare replied:
Ok I replaced the Display impl with Debug impl with https://github.com/Vector35/binaryninja-api/pull/6789/commits/90062c77233d2e9650a93b0cbc5985c84ade1c90

bdash commented on rust/src/string.rs:
```diff
@@ -87,34 +102,11 @@ impl BnString {
         res
     }
-
-    pub fn as_str(&self) -> &str {
```
I guess the idea is that these methods are available via deref to `CStr` so there's no need to wrap them?

mkrasnitski replied:
I think this is an antipattern in this case. The `Deref` implementation for `BnString` requires a linear scan, whereas for things like `String/&str` it's a constant-time operation. Deref should be reserved for smart pointers or constant-time pointer operations, and other uses are heavily warned against: https://rust-unofficial.github.io/patterns/anti_patterns/deref.html

bdash commented on rust/src/lib.rs:
```diff
@@ -299,31 +299,33 @@ pub fn set_bundled_plugin_directory(new_dir: impl AsRef<Path>) {
-    let bn_settings_file_name = unsafe { BnString::from_raw(settings_file_name_ptr) };
-    PathBuf::from(bn_settings_file_name.to_string())
+    let settings_file_path_str = unsafe { BnString::into_string(settings_file_name_ptr) };
+    PathBuf::from(settings_file_path_str)
```
File system paths are not required to be UTF-8 on all platforms. I don't know whether it matters for these sorts of paths, but it could be handled by going from a `BnString` to an `OsStr` then to a `PathBuf` rather than via `String`.

bdash commented on rust/src/section.rs:
```diff
@@ -90,12 +90,12 @@ impl Section {
 
-    pub fn name(&self) -> BnString {
-        unsafe { BnString::from_raw(BNSectionGetName(self.handle)) }
+    pub fn name(&self) -> String {
```
If sections are named with the sections taken from the loaded binary, I don't think there's any guarantee they'll be UTF-8 names.

Doing a lossy conversion here will make it so that `view.section_by_name(section.name()))` fails. That doesn't seem great.

bdash commented on rust/src/headless.rs:
```diff
@@ -71,7 +71,7 @@ pub fn init() -> Result<(), InitializationError> {
 /// ⚠️ Important! Must be called at the end of scripts. ⚠️
 pub fn shutdown() {
-    match crate::product().as_str() {
+    match crate::product().to_string().as_str() {
```
Isn't `crate::product()` already a `String`?

emesare committed:
[Rust] Simplify enterprise initialization

We don't need to introduce extra global state when the goal is to not release floating licenses in shutdown.

emesare committed:
[Rust] Simplify `BnStrCompatible` trait

Followup to https://github.com/Vector35/binaryninja-api/pull/5897/

This simplifies usage of the trait in user code, should just be able to `to_cstr` to get the cstr repr and then call `as_ptr`.

Co-authored-by: Michael Krasnitski <michael.krasnitski@gmail.com>

emesare committed:
[Rust] Simplify usage surrounding c strings

`cstring.as_ref().as_ptr() as *const c_char` -> `cstring.as_ptr()`
`cstring.as_ref().as_ptr() as *mut _` -> `cstring.as_ptr()`
`cstring.as_ptr() as *const c_char` -> `cstring.as_ptr()`

With a few fixes for cstrings that might be dropped prematurely.

emesare committed:
[Rust] Make `Section::name` return `BnString`

Section names come from the binary itself, we want to preserve the name as is

emesare committed:
[Rust] More cleanup regarding `BnString`

- Removed `to_string` shortcut from `BnString`.
- Misc formatting

emesare committed:
[Rust] Misc clippy lints

emesare commented:
Threw [this](https://github.com/Vector35/binaryninja-api/pull/6789/commits/b3aff171eec118c1517cc150837c7dbe7d001441) on the branch, still need to run more tests, I wouldn't spend time looking at it right now.

emesare commented:
CC @mkrasnitski I added you as a co-author to https://github.com/Vector35/binaryninja-api/pull/6789/commits/e66b8785c211f69e384978d1c9b5435a5e6dcbb1 seeing as it is effectively a continuation of your PR. Let me know if that is alright, and also if you have any input into this PR.

mkrasnitski commented:
Why not just merge #5897? There are no merge conflicts and then you can rebase this PR on top without the string commit.

I did a local rebase to test things, and it also let me compare your string commit with mine. They do essentially the same thing except you add a couple lines to the docstring for `BnString`. ~~However, you also left all the leftover `as *const c_char` casts which are unnecessary because `CStr::as_ptr` already returns `*const c_char`. Whereas, my PR removes them which improves formatting in a lot of places by getting rid of line breaks.~~ EDIT: I didn't see that you had that as a separate commit, my bad.

emesare commented on rust/src/architecture.rs:
```diff
@@ -1404,7 +1404,8 @@ impl CoreArchitecture {
 
     pub fn by_name(name: &str) -> Option<Self> {
-        let handle = unsafe { BNGetArchitectureByName(name.to_cstr().as_ptr() as *mut _) };
+        let name = name.to_cstr();
```
Premature drop possible

mkrasnitski commented on rust/src/string.rs:
```diff
@@ -186,110 +183,124 @@ unsafe impl CoreArrayProviderInner for BnString {
 
-    fn into_bytes_with_nul(self) -> Self::Result;
+    fn to_cstr(self) -> Self::Result;
 }
```
Having `to_cstr` take ownership can lead to some footguns around pointers to temporaries the same way `BnStrCompatible` did. In my PR, I have `as_cstr` borrow `&self` so that writing `as_cstr().as_ptr()` inline isn't  a problem except if we're dealing with iterators of string-like objects where a local collection of `&CStr` needs to be created first, and then another collection of pointers. That also lets me return a concrete `Cow<'_, CStr>` given that I have access to the lifetime `&self`, instead of requiring an associated type.

emesare replied:
The Cow lifetime does not affect how long the owned memory lives:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7924a5caad39f442aab62c6eba1f7ad1

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=e6271f359b299aa003268326766603c1

`bad` is still bad with your PR, it will free the underlying memory even if you tie the lifetime of &self to Cow.

emesare committed:
[Rust] Rename `AsCStr` to `IntoCStr`

emesare committed:
[Rust] Add string tests

emesare committed:
[Rust] Reduce usage of `IntoCStr` in function signatures

This is being done to reduce complexity in function signatures, specifically many of the strings we are passing ultimately should be new types themselves instead of "just strings", things such as type ids.

Another place which was confusing was dealing with filesystem related APIs, this commit turns most of those params into a stricter `Path` type.

This is bringing the rust api more inline with both python and C++, where the wrapper eagerly converts the string into the languages standard string type.

Special consideration must be made for symbols or other possible non utf-8 objects.

This commit will be followed up with one that adds the `IntoCStr` bound on API's we want to keep as invalid utf-8 so we can for example, get section by name on a section with invalid utf-8.

emesare committed:
[Rust] Add section tests

emesare committed:
[Rust] Retain core string for section strings

Might want to key off the section string in the core, so we should give it back as a `BnString`

emesare committed:
[Rust] Correct impls for `Section`

Previously we were comparing the section raw pointer and hashing it

emesare committed:
[Rust] Construct a `SectionBuilder` from a `Section`

emesare committed:
[Rust] Metadata API improvements

- Don't use Result for some or none values
- Fix freeing of key value store objects
- Improve ergonomics a little bit
- Add unit tests

emesare committed:
[Rust] Improve `FileAccessor`

- Add unit tests
- Expose read/write functionality
- Add some much needed documentation when passing to memory map

emesare committed:
[Rust] Add doc comments for undo actions

emesare committed:
[Rust] Misc cleanup

emesare committed:
[Rust] Flow graph API improvements

- Fixed some misc bugs
- Added real FlowGraphEdge type
- Added accessors for node edges
- Added interaction handler to the flow graph example to dump the flow graph to stdin
- Split out the flow graph API into smaller files

emesare committed:
[Rust] Fix misc typos

emesare committed:
[Rust] Support SSA form properly in low level IL

emesare committed:
[Rust] Remove unneeded mut from Architecture trait signatures

These are never expressed, just adds another unneeded constraint on the impl

emesare committed:
[Rust] Add `CoreRegister` name to the `Debug` impl

emesare committed:
[Rust] Add Assert and ForceVersion LLIL operations

emesare committed:
[Rust] Remove `NonSSAVariant` bound from `LowLevelILFunction`

We don't do enough with the lifted il != non lifted il to justify the bound.

This makes modifying IL much less work as the historical lifted il bound is gone.

emesare committed:
[Rust] Move `BinaryReader` and `BinaryWriter` into `binary_view` module

Both of these are associated directly to a `BinaryView` and only exist as accessors onto it.

emesare committed:
[Rust] Consult secrets provider when trying to connect to a remote

emesare committed:
[Rust] Misc cleanup

emesare committed:
[Rust] Interaction handler API fixes

emesare committed:
[Rust] Flowgraph API improvements

bdash committed:
[Rust] Don't panic when a Rust plug-in encounters an unhandled MLIL instruction

`MediumLevelILInstruction` does not yet handle `MLIL_CALL_OUTPUT`,
`MLIL_CALL_PARAM`, `MLIL_CALL_PARAM_SSA`, `MLIL_CALL_OUTPUT_SSA`,
`MLIL_MEMORY_INTRINSIC_OUTPUT_SSA`, or `MLIL_MEMORY_INTRINSIC_SSA`. Map
these to a `NotYetImplemented` kind rather than panicking since a panic
takes down the entire app.

emesare committed:
[Rust] Remove useless alias function

emesare committed:
[Rust] Check lifted call for params and output in mlil test

emesare committed:
[Rust] Remove extra assert in flowgraph example

emesare committed:
[Rust] Simplify operand list retrieval for MLIL and HLIL

emesare committed:
[Rust] Add missing instruction types to MLIL

emesare committed:
[Rust] Misc cleanup

emesare committed:
[Rust] Add type printer tests

emesare committed:
[Rust] Make TypeLibrary ref counted

And some other misc cleanup

emesare committed:
[Rust] Add `Display` and `Debug` impl to `Metadata`

emesare committed:
[Rust] Misc type library cleanup

emesare committed:
[Rust] Enclose the `BnString` with ""  in the `Debug` impl

emesare committed:
[Rust] Add `Platform::get_type_library_by_name`

emesare committed:
[Rust] Add type library tests

emesare committed:
[Rust] Add type library example

emesare committed:
[Rust] Keep `TypeArchiveSnapshotId` as a c_str when passing to core

emesare committed:
[Rust] Misc cleanup regarding c strings
